File: waterproof/src/main/java/mx/yellowme/waterproof/mvp/lists/ListElementsContract.java
package mx.yellowme.waterproof.mvp.lists;

import java.util.List;

import mx.yellowme.waterproof.mvp.BaseView;

public interface ListElementsContract {

    interface View<Model> extends BaseView {

        void setEmptyListState(boolean isEmpty);

        void showElements(List<Model> elements);

        void showMoreElements(List<Model> elements);

        void showElementDetail(Model elementSnap);

    }

    interface Presenter<Model> {

        void loadElements(boolean forceUpdate);

        void openElement(Model element);

    }
}

File: waterproof/src/main/java/mx/yellowme/waterproof/common/mvp/lists/ListElementsContract.kt
package mx.yellowme.waterproof.common.mvp.lists

import mx.yellowme.waterproof.common.mvp.BaseView

interface ListElementsContract {

    interface View<Model> : BaseView {

        fun empty(isEmpty: Boolean)

        fun show(elements: MutableList<Model>)

        fun showMore(elements: MutableList<Model>)

        fun showElement(snap: Model)

    }

    interface Presenter<Model> {

        fun load(forceUpdate: Boolean)

        fun open(element: Model)

        fun loadMore(currentPage: Int)

        fun search(criteria: String)
    }
}